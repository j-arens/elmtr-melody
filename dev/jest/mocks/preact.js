module.exports = {
    unmountComponent: jest.fn(),
};
