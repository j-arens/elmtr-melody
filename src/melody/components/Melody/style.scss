.Melody {
    height: 100%;

    &--fetching {
        display: none;
    }

    &--isDragging {
        cursor: move;
    }
}