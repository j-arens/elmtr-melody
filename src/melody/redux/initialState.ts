import { MachineStates } from '@state-machine/type';
import { View } from './type';

export default {
    machine: {
        currentState: 'stopped' as MachineStates,
        lastState: 'stopped' as MachineStates,
    },
    audio: {
        tracks: [],
        currentTrack: 0,
        lastTrack: 0,
        shuffle: false,
        repeat: false,
        volume: 1,
        currentTime: 0,
        timeSync: 0,
        playbackRate: 1,
        filelength: 0,
    },
    ui: {
        view: 'slider' as View,
        gliderIsDragging: false,
        volIsDragging: false,
        showDock: false,
    },
};
