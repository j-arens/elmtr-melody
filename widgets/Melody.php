<?php

namespace Melody\Widgets;

use DownShift\Container\Container;
use Elementor\Widget_Base;

class Melody extends Widget_Base implements WidgetInterface {

    /**
     * @var string
     */
    protected $handle = 'melody-audio-player';

    /**
     * @var string
     */
    protected $title = 'Melody Audio Player';

    /**
     * @var string
     */
    protected $icon = 'eicon-play';

    /**
     * @var string
     */
    protected $categories = ['general-elements'];

    /**
     * @var ViewInterface
     */
    protected $view;

    /**
     * @var array[]
     */
    protected $stacks = [];

    /**
     * @var number
     */
    protected static $instances = 0;

    /**
     * Constructor
     * 
     * @param array $data
     * @param mixed $args
     */
    public function __construct($data = [], $args = null) {
        parent::__construct($data, $args);
        $this->injectDependencies();
        add_action('elementor/frontend/after_enqueue_scripts', [$this, 'enqueueApp']);
    }

    /**
     * Elementor doesn't support dynamic dependency
     * injection of classes that extend Widget_Base
     * in the constructor, so we do it here
     */
    protected function injectDependencies() {
        $container = Container::getInstance();
        $this->view = $container->make('Melody\Core\ViewInterface');
        $this->stacks = $container['melodyControlStacks'];
    }
    
    /**
     * Get widget name
     * 
     * @return string
     */
    public function get_name() {
        return $this->handle;
    }

    /**
     * Get widget title
     * 
     * @return string
     */
    public function get_title() {
        return $this->title;
    }

    /**
     * Get widget icon
     * 
     * @return string
     */
    public function get_icon() {
        return $this->icon;
    }

    /**
     * Get widget categories
     * 
     * @return array
     */
    public function get_categories() {
        return $this->categories;
    }

    /**
     * Register controls with Elementor
     */
    protected function _register_controls() {
        if (empty($this->stacks)) {
            $this->injectDependencies();
        }

        foreach($this->stacks as $stack) {
            $this->start_controls_section(
                $stack['handle'],
                $stack['config']
            );

            foreach($stack['inputs'] as $input) {
                if (isset($input['isGroup']) && $input['isGroup']) {
                    $this->add_group_control(
                        $input['handle'],
                        $input['config']
                    );
                } else {
                    $this->add_control(
                        $input['handle'],
                        $input['config']
                    );
                }
            }

            $this->end_controls_section();
        }
    }

    /**
     * Enqueue Melody styles & scripts
     */
    public function enqueueApp() {
        wp_enqueue_style(
            'melody-app-style',
            plugins_url('temp-melody-js/style.css', STELE_MELODY_ROOT),
            null,
            filemtime(STELE_MELODY_DIR . '/temp-melody-js/style.css'),
            'all'
        );
        
        wp_enqueue_script(
            'melody-app-js',
            plugins_url('temp-melody-js/melody.bundle.js', STELE_MELODY_ROOT),
            null,
            filemtime(STELE_MELODY_DIR . '/temp-melody-js/melody.bundle.js'),
            true
        );

        wp_register_script(
            'melody-adapter-js',
            plugins_url('assets/js/adapter.bundle.js', STELE_MELODY_ROOT),
            ['elementor-frontend', 'melody-app-js'],
            filemtime(STELE_MELODY_DIR . '/assets/js/adapter.bundle.js'),
            true
        );

        wp_localize_script('melody-adapter-js', 'MELODY_ENV', [
            'pluginsUrl' => plugins_url(),
            'siteUrl' => get_site_url(),
        ]);

        wp_enqueue_script('melody-adapter-js');
    }

    /**
     * Render widget template
     */
    protected function render() {
        ++static::$instances;
        $this->view->render(STELE_MELODY_DIR . '/templates/melodyRoot.php', [
            'settings' => $this->get_settings(),
            'instance' => static::$instances,
        ]);
    }
}