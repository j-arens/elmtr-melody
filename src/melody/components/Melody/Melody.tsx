import DefaultError from '@components/Fault/DefaultError';
import FaultContainer from '@components/Fault/FaultContainer';
import NoTracksEditor from '@components/Fault/NoTracksEditor';
import NoTracksFront from '@components/Fault/NoTracksFront';
import ShapeShifter from '@components/ShapeShifter/';
import { GLOBAL } from '@constants';
import { Track } from '@redux/type';
import { Action } from '@redux/type';
import { MachineAction, MachineStates } from '@state-machine/type';
import { getRandomNumInRange, timeout } from '@utils/index';
import * as classnames from 'classnames';
import { Component, h } from 'preact';
const s = require('./style.scss');

export interface StateProps {
    currentState: MachineStates;
    lastState: MachineStates;
    tracks: Track[];
    currentTrack: number;
    currentTime: number;
    gliderIsDragging: boolean;
    volIsDragging: boolean;
    volume: number;
    timeSync: number;
    repeat: boolean;
    playbackRate: number;
}

export interface DispatchProps {
    cycleState: (action: MachineAction) => Action;
    updateCurrentTime: (nextTime: number) => Action;
    nextTrack: () => Action;
    setFilelength: (length: number) => Action;
}

type Props = StateProps & DispatchProps;

export default class extends Component<Props, {}> {
    AudioInterface: HTMLAudioElement = new Audio();

    componentDidMount() {
        this.bindAudioInterfaceEvents();
        this.setInterfaceSource(this.props);
    }

    componentWillUnmount() {
        this.removeAudioInterfaceEvents();
    }

    componentDidUpdate(prevProps: Props) {
        const {
            currentState,
            volume,
            tracks,
            currentTrack,
            timeSync,
            repeat,
            playbackRate,
            currentTime,
        } = this.props;

        const stateChanged = prevProps.currentState !== currentState;
        const volumeChanged = prevProps.volume !== volume;
        const tracksChanged = prevProps.tracks !== tracks;
        const currentTrackChanged = prevProps.currentTrack !== currentTrack;
        const timeSyncChanged = prevProps.timeSync !== timeSync;
        const repeatChanged = prevProps.repeat !== repeat;
        const playbackRateChanged = prevProps.playbackRate !== playbackRate;

        if (stateChanged) {
            this.updateInterfaceState(currentState);
        }

        if (tracksChanged || currentTrackChanged) {
            this.setInterfaceSource(this.props);
        }

        if (volumeChanged) {
            this.setInterfaceVolume(volume);
        }

        if (timeSyncChanged) {
            this.setInterfaceTime(currentTime);
        }

        if (repeatChanged) {
            this.setInterfaceRepeat(repeat);
        }

        if (playbackRateChanged) {
            this.setInterfacePlaybackRate(playbackRate);
        }
    }

    bindAudioInterfaceEvents() {
        this.AudioInterface.addEventListener('waiting', this.onInterfaceBuffering);
        this.AudioInterface.addEventListener('canplay', this.onCanPlay);
        this.AudioInterface.addEventListener('timeupdate', this.onTimeUpdate);
        this.AudioInterface.addEventListener('ended', this.onEnded);
        this.AudioInterface.addEventListener('loadedmetadata', this.onLoadedMeta);
    }

    removeAudioInterfaceEvents() {
        this.AudioInterface.removeEventListener('waiting', this.onInterfaceBuffering);
        this.AudioInterface.removeEventListener('canplay', this.onCanPlay);
        this.AudioInterface.removeEventListener('timeupdate', this.onTimeUpdate);
        this.AudioInterface.removeEventListener('ended', this.onEnded);
        this.AudioInterface.removeEventListener('loadedmetadata', this.onLoadedMeta);
    }

    updateInterfaceState(nextState: MachineStates) {
        switch (nextState) {
            case 'fetching':
            case 'stopped':
            case 'fault': {
                this.AudioInterface.pause();
                break;
            }
            case 'playing': {
                this.AudioInterface.play().catch(this.onPlayError);
                break;
            }
            case 'buffering': {
                break;
            }
            default: {
                this.AudioInterface.pause();
                break;
            }
        }
    }

    async setInterfaceSource({ currentState, currentTrack, tracks } = this.props) {
        if (!tracks.length) {
            this.AudioInterface.src = '';
            return;
        }

        const { cycleState } = this.props;
        const nextSource = tracks[currentTrack].source_url;

        if (this.AudioInterface.src === nextSource) {
            return;
        }

        const load = (source) => {
            this.AudioInterface.src = source;
            this.AudioInterface.load();
        };

        if (currentState === 'playing') {
            await cycleState('STOP');
            load(nextSource);
            cycleState('PLAY');
            return;
        }

        load(nextSource);
    }

    setInterfaceTime(nextTime: number) {
        this.AudioInterface.currentTime = nextTime;
    }

    setInterfaceVolume(newVolume: number) {
        this.AudioInterface.volume = newVolume;
    }

    setInterfaceRepeat(repeat: boolean) {
        this.AudioInterface.loop = repeat;
    }

    setInterfacePlaybackRate(rate: number) {
        this.AudioInterface.playbackRate = rate;
    }

    onInterfaceBuffering = () => {
        const { cycleState, currentState } = this.props;

        if (currentState === 'buffering') {
            return;
        }

        const clear = timeout(10000, () => {
            if (this.props.currentState === 'buffering') {
                cycleState('FAILED');
            } else {
                clear();
            }
        });

        cycleState('LOADING');
    }

    onCanPlay = () => {
        const { currentState, lastState, cycleState } = this.props;

        if (currentState !== 'buffering') {
            return;
        }

        if (lastState === 'playing') {
            cycleState('PROCEED');
            return;
        }

        cycleState('READY');
    }

    onPlayError = (e) => {
        const { cycleState } = this.props;
        if (e.name && e.name !== 'AbortError') {
            cycleState('FAILED');
        }
    }

    onTimeUpdate = () => {
        const nextTime = Number(this.AudioInterface.currentTime.toFixed(0));
        const { currentTime, updateCurrentTime, gliderIsDragging } = this.props;

        if (nextTime === currentTime || gliderIsDragging) {
            return;
        }

        updateCurrentTime(nextTime);
    }

    onEnded = () => {
        const { nextTrack, repeat } = this.props;

        if (repeat) {
            return;
        }

        nextTrack();
    }

    onLoadedMeta = () => {
        const { setFilelength } = this.props;
        const { duration } = this.AudioInterface;
        setFilelength(duration);
    }

    getMachineAction() {
        const { currentState } = this.props;
    }

    render({ currentState, gliderIsDragging, volIsDragging, tracks }: Props) {
        const classes = classnames(
            s.Melody,
            s[`Melody--${currentState}`],
            { [s['Melody--isDragging']]: gliderIsDragging || volIsDragging },
        );

        if (currentState !== 'fetching' && !tracks.length) {
            if (GLOBAL.elementorFrontend && GLOBAL.elementorFrontend.isEditMode()) {
                return (
                    <FaultContainer allowReload={false}>
                        <NoTracksEditor />
                    </FaultContainer>
                );
            }
            return (
                <FaultContainer allowReload={false}>
                    <NoTracksFront />
                </FaultContainer>
            );
        }

        if (currentState === 'fault') {
            return (
                <FaultContainer allowReload>
                    <DefaultError />
                </FaultContainer>
            );
        }

        return (
            <div class={classes} data-melody-border>
                <ShapeShifter />
            </div>
        );
    }
}
