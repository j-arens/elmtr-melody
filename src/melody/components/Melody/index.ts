import {
    cycleState,
    nextTrack,
    setFilelength,
    updateCurrentTime,
} from '@redux/actions';
import { State } from '@redux/type';
import { MachineAction } from '@state-machine/type';
import { connect } from 'preact-redux';
import Melody, { DispatchProps, StateProps } from './Melody';

const mapStateToProps = (state: State) => ({
    currentState: state.currentState,
    lastState: state.lastState,
    tracks: state.tracks,
    currentTrack: state.currentTrack,
    currentTime: state.currentTime,
    gliderIsDragging: state.ui.gliderIsDragging,
    volIsDragging: state.ui.volIsDragging,
    volume: state.volume,
    timeSync: state.timeSync,
    repeat: state.repeat,
    playbackRate: state.playbackRate,
});

const mapDispatchToProps = (dispatch) => ({
    cycleState: (action: MachineAction) => dispatch(cycleState(action)),
    updateCurrentTime: (nextTime: number) => dispatch(updateCurrentTime(nextTime)),
    nextTrack: () => dispatch(nextTrack()),
    setFilelength: (length: number) => dispatch(setFilelength(length)),
});

export default connect<
    StateProps,
    DispatchProps,
    {}
>(mapStateToProps, mapDispatchToProps)(Melody);
