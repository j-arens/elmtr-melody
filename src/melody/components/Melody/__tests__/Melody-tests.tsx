import configureStore from '@redux/index';
import initialState from '@redux/initialState';
import { h } from 'preact';
import { Provider } from 'preact-redux';
import { render } from 'preact-render-to-string';
const track = require('@fixtures/track');
import Melody from '../index';

describe('Melody', () => {
    let state;

    beforeEach(() => {
        state = initialState;
    });

    it('renders an error if there arent any tracks', () => {
        const component = render(
            <Provider store={configureStore(state)}>
                <Melody />
            </Provider>,
        );
        expect(component).toMatchSnapshot();
    });

    it('renders an editor specific error message if there arent any tracks', () => {
        global.elementorFrontend = {
            isEditMode: jest.fn().mockReturnValue(true),
        };
        const store = configureStore(state);
        const component = render(
            <Provider store={store}>
                <Melody />
            </Provider>,
        );
        expect(component).toMatchSnapshot();
    });

    it('renders an error if the currentState is fault', () => {
        const testState = {
            ...state,
            state: {
                ...state.state,
                currentState: 'fault',
            },
            audio: {
                ...state.audio,
                tracks: [{}],
            },
        };
        const store = configureStore(testState);
        const component = render(
            <Provider store={store}>
                <Melody />
            </Provider>,
        );
        expect(component).toMatchSnapshot();
    });

    it('flourishes into a nice little application if the conditions are right', () => {
        const testState = {
            ...state,
            audio: {
                ...state.audio,
                tracks: [track],
            },
        };
        const store = configureStore(testState);
        const component = render(
            <Provider store={store}>
                <Melody />
            </Provider>,
        );
        expect(component).toMatchSnapshot();
    });
});
