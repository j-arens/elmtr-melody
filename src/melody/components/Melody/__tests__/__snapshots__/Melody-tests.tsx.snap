// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Melody flourishes into a nice little application if the conditions are right 1`] = `
<div class="Melody Melody--stopped">
  <div class="melody__viewContainer" data-melody-view="slider">
    <div class="preview melody-c-preview-bg melody-c-preview-min-height melody-c-preview-padding">
      <ul class="Slideshow preview__slideshow melody-c-slider-transition melody-c-slider-timing"
      style="transform: translate3d(0%, 0px, 0px);">
        <li class="Slideshow__slide">
          <figure class="Slideshow__image melody-c-artwork-size melody-c-artwork-repeat melody-c-artwork-position melody-c-artwork-attachment"
          style="background-image: url();"></figure>
        </li>
      </ul>
      <div class="preview__header">
        <div class="melody-c-vol-primary-color melody-c-vol-hover-color melody-c-vol-width volumeCtrl">
          <svg class="volumeCtrl__icon elementor-tab-title" xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24">
            <circle class="volumeCtrl__levelOutline" cx="12" cy="12" r="12"></circle>
            <circle class="volumeCtrl__level" style="stroke-dashoffset: 0;"
            cx="12" cy="12" r="12"></circle>
            <path d="M3 9v6h4l5 5V4L7 9H3zm13.5 3c0-1.77-1.02-3.29-2.5-4.03v8.05c1.48-.73 2.5-2.25 2.5-4.02zM14 3.23v2.06c2.89.86 5 3.54 5 6.71s-2.11 5.85-5 6.71v2.06c4.01-.91 7-4.49 7-8.77s-2.99-7.86-7-8.77z"></path>
          </svg>
        </div>
        <div class="dock ">
          <button class="btn dock__toggle " role="button">
            <svg class="dock__toggleIcon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
              <path d="M19.43 12.98c.04-.32.07-.64.07-.98s-.03-.66-.07-.98l2.11-1.65c.19-.15.24-.42.12-.64l-2-3.46c-.12-.22-.39-.3-.61-.22l-2.49 1c-.52-.4-1.08-.73-1.69-.98l-.38-2.65C14.46 2.18 14.25 2 14 2h-4c-.25 0-.46.18-.49.42l-.38 2.65c-.61.25-1.17.59-1.69.98l-2.49-1c-.23-.09-.49 0-.61.22l-2 3.46c-.13.22-.07.49.12.64l2.11 1.65c-.04.32-.07.65-.07.98s.03.66.07.98l-2.11 1.65c-.19.15-.24.42-.12.64l2 3.46c.12.22.39.3.61.22l2.49-1c.52.4 1.08.73 1.69.98l.38 2.65c.03.24.24.42.49.42h4c.25 0 .46-.18.49-.42l.38-2.65c.61-.25 1.17-.59 1.69-.98l2.49 1c.23.09.49 0 .61-.22l2-3.46c.12-.22.07-.49-.12-.64l-2.11-1.65zM12 15.5c-1.93 0-3.5-1.57-3.5-3.5s1.57-3.5 3.5-3.5 3.5 1.57 3.5 3.5-1.57 3.5-3.5 3.5z"></path>
            </svg>
          </button>
        </div>
      </div>
      <div class="preview__marquee">
        <p class="preview__title melody-c-title-font melody-c-title-shadow melody-c-title-color">lol-title</p>
        <p class="preview__artist melody-c-artist-font melody-c-artist-shadow melody-c-artist-color">lol-artist</p>
      </div>
      <div class="preview__glider">
        <div class="glider elementor-tab-title">
          <div class="glider__playback">
            <div class="glider__trackbar" style="margin-left: 0%;" data-melody-scrubber-progress>
              <div class="glider__handle" data-melody-scrubber-handle></div>
            </div>
          </div>
        </div>
      </div>
      <div class="preview__footer">
        <span class="preview__time melody-c-time-font melody-c-time-shadow melody-c-time-color">0:00</span>
        <span class="preview__time melody-c-time-font melody-c-time-shadow melody-c-time-color">0:00</span>
      </div>
    </div>
    <div class="controlbar melody-c-controlbar melody-control-bar">
      <button class="btn melody-c-btn-primary-color melody-c-btn-hover-color melody-c-btn-space playbackCtrl melody-playbackCtrl melody-shuffle"
      role="button">
        <svg class="playbackCtrl__icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path d="M7 7h10v3l4-4-4-4v3H5v6h2V7zm10 10H7v-3l-4 4 4 4v-3h12v-6h-2v4z"></path>
        </svg>
      </button>
      <button class="btn melody-c-btn-primary-color melody-c-btn-hover-color melody-c-btn-space playbackCtrl melody-playbackCtrl playbackCtrl--disabled"
      role="button">
        <svg class="playbackCtrl__icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path d="M11 18V6l-8.5 6 8.5 6zm.5-6l8.5 6V6l-8.5 6z"></path>
        </svg>
      </button>
      <button data-playbackRate class="btn melody-c-btn-primary-color melody-c-btn-hover-color melody-c-btn-space playbackCtrl melody-playbackCtrl melody-controlBar__catalyst"
      role="button">
        <svg class="playbackCtrl__icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <g>
            <circle class="melody-playbackCtrl__ring" cx="12" cy="12" r="12"></circle>
            <path d="M8 5v14l11-7z"></path>
          </g>
        </svg>
      </button>
      <button class="btn melody-c-btn-primary-color melody-c-btn-hover-color melody-c-btn-space playbackCtrl melody-playbackCtrl playbackCtrl--disabled"
      role="button">
        <svg class="playbackCtrl__icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path d="M4 18l8.5-6L4 6v12zm9-12v12l8.5-6L13 6z"></path>
        </svg>
      </button>
      <button class="btn melody-c-btn-primary-color melody-c-btn-hover-color melody-c-btn-space playbackCtrl melody-playbackCtrl melody-repeat"
      role="button">
        <svg class="playbackCtrl__icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
          <path d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"></path>
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Melody renders an editor specific error message if there arent any tracks 1`] = `
<div class="errorContainer">
  <p class="melody__body">No tracks loaded, click here to get started.</p>
</div>
`;

exports[`Melody renders an error if the currentState is fault 1`] = `
<div class="errorContainer">
  <span>
    <svg class="errorContainer__icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
      <path d="m110 150 20 0 0 20-20 0zm0-80 20 0 0 60-20 0zm9.9-50C64.7 20 20 64.8 20 120 20 175.2 64.7 220 119.9 220 175.2 220 220 175.2 220 120 220 64.8 175.2 20 119.9 20Zm0.1 180c-44.2 0-80-35.8-80-80 0-44.2 35.8-80 80-80 44.2 0 80 35.8 80 80 0 44.2-35.8 80-80 80z"></path>
    </svg>
    <p class="melody__body">Something went wrong!</p>
  </span>
  <button class="btn errorContainer__reloadBtn" role="button">Reload</button>
</div>
`;

exports[`Melody renders an error if there arent any tracks 1`] = `
<div class="errorContainer">
  <p class="melody__body">Doh! It looks like there aren't any tracks loaded.</p>
</div>
`;
